common/kotlinx-coroutines-core-common/src/channels/Channel.kt

```diff
@@ -268,6 +301,75 @@ public interface ReceiveChannel<out E> {
     public fun cancel(cause: Throwable? = null): Boolean
 }
 
+/**
+ * A discriminated union of [ReceiveChannel.receiveOrClosed] result,
+ * that encapsulates either successfully received element of type [T] from the channel or a close cause.
+ */
+@Suppress("NON_PUBLIC_PRIMARY_CONSTRUCTOR_OF_INLINE_CLASS")
+public inline class ValueOrClosed<out T>
+internal constructor(private val holder: Any?) {
+    /**
+     * Returns `true` if this instance represents received element.
+     * In this case [isClosed] returns `false`.
+     */
+    public val isValue: Boolean get() = holder !is Closed
+
+    /**
+     * Returns `true` if this instance represents close cause.
+     * In this case [isValue] returns `false`.
+     */
+    public val isClosed: Boolean get() = holder is Closed
+
+    /**
+     * Returns received value if this instance represents received value or throws [IllegalStateException] otherwise.
+     */
+    @Suppress("UNCHECKED_CAST")
+    public val value: T
```

Comment:
So we have the following methods on `ValueOrClosed`:
* `value`
* `valueOrNull`
* `valurOrThrow`

Which is kind of confusing, since the channel has the following `receive` variants:

* `receive` which corresponds to `receiveOrClosed().valueOrThrow`... huh?
* `receiveOrNull` which corresponds to `receiveOrClosed().valueOrNull`
*  `receiveOrClosed`

For the sake of consistency I'd suggest to drop `value` and rename `valueOrThrow` to simply `value`.

Reply:
I've moved this comment to the source in #1343 for future decision.